test/hooks.on-listen.test.js

```diff
@@ -0,0 +1,614 @@
+      t.equal(order++, 1, 'called in childOne')
+      t.equal(this.pluginName, childOne.pluginName, 'the this binding is the right instance')
+      done()
+    })
+    childOne.register(async (childTwo, o) => {
+      childTwo.addHook('onListen', async function () {
+        t.equal(order++, 2, 'called in childTwo')
+        t.equal(this.pluginName, childTwo.pluginName, 'the this binding is the right instance')
+      })
+    })
+  })
+  fastify.listen({ port: 0 }, (err) => {
+    t.equal(fastify.server.address().address, localhost)
+    t.error(err)
+  })
+})
+t.test('onListen nonlocalhost should work in order with callback in sync', t => {
+  t.plan(4)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, '1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 1, '2nd called in root')
+    done()
+  })
+
+  fastify.listen({ host: '::1', port: 0 }, (err) => {
+    t.equal(fastify.server.address().address, '::1')
+    t.error(err)
+  })
+})
+t.test('onListen nonlocalhost should work in order with callback in async', t => {
+  t.plan(4)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, '1st called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 1, '2nd called in root')
+  })
+
+  fastify.listen({ host: '::1', port: 0 }, (err) => {
+    t.equal(fastify.server.address().address, '::1')
+    t.error(err)
+  })
+})
+t.test('onListen nonlocalhost with callback should manage error in sync', t => {
+  t.plan(5)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', function (done) {
+    t.pass('1st called in root')
+    done()
+  })
+
+  fastify.addHook('onListen', function (done) {
+    t.equal(order++, 0, 'error sync called in root')
+    done(new Error('FAIL ON LISTEN'))
+  })
+
+  fastify.listen({ host: '::1', port: 0 }, (err) => {
+    if (err) {
+      t.ok(err, 'FAIL ON LISTEN')
+    } else {
+      t.equal(fastify.server.address().address, '::1')
+      t.error(err)
+    }
+  })
+})
+t.test('onListen nonlocalhost with callback should manage error in async', t => {
+  t.plan(5)
+  const fastify = Fastify()
+  t.teardown(fastify.close.bind(fastify))
+  let order = 0
+
+  fastify.addHook('onListen', async function () {
+    t.pass('1st called in root')
+  })
+
+  fastify.addHook('onListen', async function () {
+    t.equal(order++, 0, 'error sync called in root')
+    throw new Error('FAIL ON LISTEN')
+  })
+
+  fastify.listen({ host: '::1', port: 0 }, (err) => {
+    if (err) {
+      t.ok(err, 'FAIL ON LISTEN')
+    } else {
+      t.equal(fastify.server.address().address, '::1')
+      t.error(err)
+    }
```

Comment:
Why the `if/else`, I assume by the test that we are testing for a error properly thrown 🤔

Reply:
You're right. We don't need that and will make the necessary changes.

Reply:
Probably should remove this logic out of the other tests as well right?
For example all callback tests have this: 
```
  fastify.listen({ host: 'localhost', port: 0 }, (err) => {
    t.equal(fastify.server.address().address, localhost)
    t.error(err)
  })
```
Maybe it should be replaced by something like this?
```
  fastify.listen({ port: 0 }, (err) => {
    if (err) {
      t.fail('Uncaught onListen error')
    } else {
      t.ok('No error')
    }
  })
```

Reply:
Depends on the test, if you are testing that the error throw, it is enough by just doing

```js
t.equal(err, 'Your error message');
```

If you are testing an error is not thrown, the following works
```js
t.error(err);
t.ok(!err);
```

Reply:
If you dont(!) expect an error, do `t.error(err)`
https://node-tap.org/docs/api/asserts/#terrorobj-message-extra